File: python-version/LC3097/minimum_subarray_length.py
import math

class Solution:
    def minimumSubarrayLength(self, nums: list[int], k: int) -> int:
        ans, n = math.inf, len(nums)
        left = 0
        bits = [0 for _ in range(32)]

        for right in range(n):
            for i in range(32):
                bits[i] += nums[right] >> i & 1
            while left <= right and self.valueOf(bits) >= k:
                ans = min(ans, right - left + 1)
                for i in range(32):
                    bits[i] -= nums[left] >> i & 1
                left += 1
        
        return -1 if ans == math.inf else ans
    
    def valueOf(self, bits: list[int]) -> int:
        n = len(bits)
        ans = 0
        for i in range(n):
            if bits[i] > 0:
                ans |= 1 << i
        return ans

File: go-version/leetcode/LC3097/minimum_subarray_length.go
package lc3097

import "math"

func minimumSubarrayLength(nums []int, k int) int {
	ans := math.MaxInt
	n := len(nums)
	bits := make([]int, 32)

	for left, right := 0, 0; right < n; right++ {
		for i := 0; i < 32; i++ {
			bits[i] += nums[right] >> i & 1
		}

		for left <= right && valueOf(bits) >= k {
			ans = min(ans, right-left+1)
			for i := 0; i < 32; i++ {
				bits[i] -= nums[left] >> i & 1
			}
			left++
		}
	}

	if ans == math.MaxInt {
		ans = -1
	}
	return ans
}

func valueOf(bits []int) int {
	ans := 0
	n := len(bits)
	for i := 0; i < n; i++ {
		if bits[i] > 0 {
			ans |= 1 << i
		}
	}
	return ans
}
